### ISSUE TITLE
Fix colors for OCaml syntax highlighting

### ISSUE THREAD

**Issue by freebroccolo** [reasonml-editor/vscode-reasonml#61]
Some of the default scopes have probably changed since I last tested.

- [x] fix constructors
- [ ] fix function and module names
- [x] fix record fields

Here is what it should look like with the Dark+ theme:

![ocaml](https://cloud.githubusercontent.com/assets/151197/26514007/20f3317c-422c-11e7-9937-9cadaa39ba46.png)

**Comment by freebroccolo** [comment]
Blocked on https://github.com/Microsoft/vscode/issues/3948.

**Comment by Khady** [comment]
The colors are not working properly when I'm using the lwt ppx. The `let` and `match` colors seems to break.

![screenshot](https://cloud.githubusercontent.com/assets/974142/26536899/9219c3b6-442a-11e7-8267-0894daaf6454.png)

```ocaml
let test () =
    let _a = 3 in
    match Some 1 with
    | Some _ -> Lwt.return_unit
    | None ->
    let%lwt _v = Lwt.return_none in
    let _placeholder = 3 in
    match Some 2 with
    | Some _ -> Lwt.return_unit
    | None -> Lwt.return_unit
```

**Comment by freebroccolo** [comment]
@Khady thanks for the report. I don't think I added any parsing for ppx stuff yet but it shouldn't be hard to fix. I'll make sure it's fixed in the upcoming released.

**Comment by freebroccolo** [comment]
@Khady I think your ppx example should be fixed on `master` now.

**Comment by Khady** [comment]
Indeed it works. Thanks!

**Comment by Khady** [comment]
One more error because of a `%` I think:

```
let test_lwt () =
  let%lwt _value = begin Lwt.return_unit end [%finally Lwt.return_unit] in
  let val2 = 3 in
  Lwt.return val2

let test f = f (1, 2)
```

gives me

![image](https://user-images.githubusercontent.com/974142/27071737-9585a238-500d-11e7-811f-a6e1d3c86655.png)

**Comment by freebroccolo** [comment]
This issue appears to be due to the `[%finally Lwt.return_unit]` part of the syntax. If it is removed it looks like it's highlighting correctly:

![opt](https://user-images.githubusercontent.com/151197/30143999-d76a8b16-9345-11e7-9062-1c97a026e026.png)

**Comment by freebroccolo** [comment]
@Khady I think your second example should be fixed now.

**Comment by Khady** [comment]
Indeed it works. Thanks!

**Comment by Enzojz** [comment]
I have string systematically un-highlighted, any explain to this?

This is what I got with old OCaml extension:

![default](https://user-images.githubusercontent.com/8238865/31827538-53c07fa6-b5b8-11e7-9fbc-b6bd0db68878.JPG)

And this is with extension:

![2](https://user-images.githubusercontent.com/8238865/31827555-608921f2-b5b8-11e7-8f7c-df7a0e13ace4.JPG)

**Comment by maelvalais** [comment]
Hi,
I have the same issue as @Enzojz: the syntax coloring stops somewhere in the document and all strings are not detected.
In the dev tools console, I get the error

     [1] - Grammar is in an endless loop - Grammar pushed & popped a rule without advancing

with the file `foo.ml`:

```ocaml
let foo ast :ast =
  let bar = ""
  in
  match ast_without_loc ast with
  | Touist_code (formulas) ->
    eval_ast_formula env (process_formulas (affect_vars formulas))
  | e -> raise_with_loc ast ("this " ^ string_of_ast e ^"\n")
```
![capture d ecran 2017-10-22 a 20 50 18](https://user-images.githubusercontent.com/2195781/31865185-b92cf9e2-b76a-11e7-9b25-f374fd007620.png)
But adding a space after `:`:
![capture d ecran 2017-10-22 a 20 50 33](https://user-images.githubusercontent.com/2195781/31865188-baccadce-b76a-11e7-8b24-e079aadec4b3.png)

----

Note: to debug tmlanguages, I hacked `grammar-debug`; to use it

    npm install -g https://github.com/maelvalais/grammar-debug#cli-in-path
    grammar-debug ~/.vscode/extensions/vscode-reasonml/syntaxes/ocaml.json foo.ml

Here is the result:

```
TOKENIZING LINE 1
let foo ast :ast =
[1] - Grammar is in an endless loop - Grammar pushed & popped a rule without advancing

  LINE CONTAINS 8 TOKENS:
    => TOKEN (0-3):
      let
      * source.ocaml
      * keyword.control markup.underline
    => TOKEN (3-4):

      * source.ocaml
    => TOKEN (4-7):
      foo
      * source.ocaml
      * entity.name.function strong emphasis
    => TOKEN (7-8):

      * source.ocaml
    => TOKEN (8-11):
      ast
      * source.ocaml
      * string.other.link variable.language variable.parameter emphasis
    => TOKEN (11-12):

      * source.ocaml
    => TOKEN (12-13):
      :
      * source.ocaml
      * variable.other.class.js message.error variable.interpolation string.regexp strong
    => TOKEN (13-19):
      ast =
      * source.ocaml

  LINE END RULE STACK CONTAINS 4 RULES:
      * [306,1] BeginEndRule "null", entered @13
      * [304,2] BeginEndRule "null", entered @13
      * [263,3] BeginEndRule "null", entered @0
      * [1,4] IncludeOnlyRule "source.ocaml", entered @-1
```

**Comment by maelvalais** [comment]
By the way, how do you do to add/edit matches in `ocaml.json` in this repo? After un-minifying it, I feel a bit lost with the ~2800 lines.

Also, when using `Developer: inspect TM scopes`, I saw that the "names" fields always contain multiple names (à la HTML classes), e.g. in `commentBlock`:
```json
"name": "comment constant.regexp",
"contentName": "emphasis"
```
How do you manage the names (HTML classes-like?) Maybe it is customary in VSCode tmLanguages? (I was used to the textmate [naming conventions](https://manual.macromates.com/en/language_grammars#naming_conventions.html)).

cc @freebroccolo

**Comment by freebroccolo** [comment]
> I have strings systematically un-highlighted, any explain to this?

It is just a scoping issue. Should be fixed in the upcoming release.

**Comment by freebroccolo** [comment]
> In the dev tools console, I get the error
>  [1] - Grammar is in an endless loop - Grammar pushed & popped a rule without advancing

I'll take a look at this. Not sure what the issue is yet.

> By the way, how do you do to add/edit matches in ocaml.json in this repo? After un-minifying it, I feel a bit lost with the ~2800 lines ☺️

The textmate grammar is generated from [this](https://github.com/reasonml-editor/vscode-reasonml/blob/master/src/syntaxes/ocaml.ts).

> How do you manage the names (CSS classes-like?) Maybe it is customary in VSCode tmLanguages? (I was used to the textmate naming conventions).

Most of the scoping stuff is [here](https://github.com/reasonml-editor/vscode-reasonml/blob/master/src/syntaxes/basis.ts). The problem with the textmate "naming conventions" is that they aren't really standard and were simply never adopted very consistently across the different themes and editors. Because of that, I don't really bother even trying to use scope names according to their "semantic" notion. Instead, I try to tune the scopes to look reasonably good across the editors and themes that I test with regularly.

**Comment by maelvalais** [comment]
Thank you for your answers!!

True, the textmate "semantic" conventions are not really well followed (and also kinf of hard to follow). That said, I noticed that the comments are colored differently than for other syntaxes (in Dark+, OCaml comments use the coloring of `constant.regexp` but in Reason comments they are colored using `comment.block`:

![capture d ecran 2017-10-22 a 22 05 24](https://user-images.githubusercontent.com/2195781/31865828-6425461a-b775-11e7-89f3-52da83871c57.png)

Maybe we could unify that?

Anyway, thank you for your great work! 😃

**Comment by freebroccolo** [comment]
The plan is to eventually unify the coloring. What happened is that I rewrote the OCaml highlighter using this more advanced approach but haven't done the same thing for the Reason highlighter yet (which was written by hand and is harder to maintain). I will probably wait until the next round of syntax changes upstream before I finish that.

**Comment by freebroccolo** [comment]
> With no space between : and ast, the highlighting breaks:
> 
> But adding a space after ::
> 
> In the dev tools console, I get the error
>  [1] - Grammar is in an endless loop - Grammar pushed & popped a rule without advancing

This should be fixed now in `master`.

**Comment by maelvalais** [comment]
Wow, that was fast! Thank you!

**Comment by maelvalais** [comment]
I went through the `ocaml.ts`. The idea of "abstracting" the grammar using typescript seems promising!

I have some questions though:
- don't you think using CSS/styling inside the syntax definition file is going to make the syntax coloring inconsistent between two languages (using the same theme)?
- maybe the fine-grained highlighting (like `let` being underlined) should be inside the theme itself?

Imagine that `let` has the name `keyword.other.function-definition.ocaml` (example taken from `vscode-ocaml`). Most themes will highlight `let` using `keyword.other`; but if you want to have underlining for this specific case in Dark+, we can add a rule in the theme that uses `keyword.other.function-definition`?

Now that I think of that, these [naming conventions](https://manual.macromates.com/en/language_grammars#naming_conventions.html) are really limiting because these "conventions" only covers the basic language elements; beyond these, everyone has his own recipe 😞 making themes and syntax definition less and less compatible with each other
